src/freenet/client/async/SplitFileInserter.java

```diff
@@ -112,9 +112,13 @@ public void cancel(ClientContext context) {
     public void schedule(ClientContext context) throws InsertException {
         cb.onBlockSetFinished(this, context);
         storage.start();
+        SplitFileInserterSender s;
+        synchronized(this) {
+            s = sender;
+        }
         if(!ctx.getCHKOnly) {
-            sender.clearWakeupTime(context);
-            sender.schedule(context);
+            s.clearWakeupTime(context);
+            s.schedule(context);
```

Comment:
The lock here isn't really necessary at all, it's mainly to shut up code analysis tools. sender is only set when we create a splitfile fetcher or when we load one from disk. Maybe we could make it an AtomicObject or volatile?

Both of the methods called are indeed thread-safe, as is more or less everything in the client layer. In detail:

clearWakeupTime(): If we don't have an RGA then we're not in the queue, which means we are not on the queue i.e. we are not registered. If so, then the schedule() will fix this. If we are on the queue, then RGA.clearWakeupTime() takes the lock on the ClientRequestSelector (i.e. the entire request queue for the given type of requests e.g. bulk CHK inserts), marks the request as sendable, and if necessary clears the cooldown for parents too (i.e. if a branch of the tree contains a sendable request, then that branch is itself sendable). In the event of a race condition (the RGA is no longer in the tree), we propagate this upwards anyway, which means that branch of the tree will be tried even though it doesn't have any requests to send; this is fine, it will be corrected automatically by the next attempt to find a request to send.

schedule() is very similar to schedule() on other requests, and is thread safe. If there is an RGA then we are already scheduled, that is we are on the request queue, so we don't need to be scheduled again. If another thread is unregistering it there are 2 possibilities:
1. It is being explicitly unregistered, by the means given in this patch. If so, it's finished, we don't need it to be scheduled. (If we restart the insert the ClientPutter will create a new SplitFileInserter).
2. It is removed from the RGA because it is not sendable. IIRC this only happens if the request is actually cancelled i.e. it returns -1 from getWakeupTime(), in which case there is a race condition but removing it from the queue is correct. This can happen if the user cancels the insert while it is still encoding, for example. If getWakeupTime() returns a positive value then it is kept on the queue, but marked as not fetchable until some point in the future; clearWakeupTime() clears this e.g. when requests fail and become retryable.